xla/tests/collective_ops_e2e_test.cc

```diff
@@ -1340,6 +1341,290 @@ TEST_F(CollectiveOpsTestE2E, NoAsyncCollectives) {
+    // from the HLO string.
+    std::string hlo_text_ref = hlo_text;
+    RE2::GlobalReplace(&hlo_text_ref, R"(, sharding=\{devices=\[[0-9,]*\].*\})",
+                       "");
+    RE2::GlobalReplace(&hlo_text_ref, R"(, sharding=\{replicated\})", "");
+
+    HloModuleConfig ref_config = GetModuleConfigForTest(/*replica_count=*/1);
+    auto ref_opts = GetDebugOptionsForTest();
+    ref_opts.set_xla_gpu_enable_triton_gemm(false);
+    ref_config.set_debug_options(ref_opts);
+    ref_config.set_num_partitions(1);
+    TF_ASSIGN_OR_RETURN(auto ref_module,
+                        ParseAndReturnVerifiedModule(hlo_text_ref, ref_config));
+    const int64_t num_params =
+        ref_module->entry_computation()->num_parameters();
+
+    auto fake_args = xla::MakeFakeArguments(ref_module.get()).value();
+    std::vector<Literal*> ref_fake_ptrs(num_params);
+    for (int i = 0; i < num_params; ++i) {
+      ref_fake_ptrs[i] = &fake_args[i];
+    }
+
+    DeviceAssignment ref_assn(/*replica_count=*/1,
+                              /*computation_count=*/1);
+    ref_assn(0, 0) = 0;
+    return HloTestBase::ExecuteReplicated(std::move(ref_module), ref_fake_ptrs,
+                                          /*num_replicas=*/1, &ref_assn,
+                                          /*run_hlo_passes=*/true,
+                                          /*use-threads=*/true);
+  }
+
+  // Execute the sharded case.
+  absl::StatusOr<std::vector<Literal>> ExecuteSharded(
+      const std::string& hlo_text, int64_t num_partitions,
+      int64_t num_replicas) {
+    HloModuleConfig config =
+        GetModuleConfigForTest(/*replica_count=*/num_replicas);
+    auto opts = GetDebugOptionsForTest();
+    opts.set_xla_gpu_enable_triton_gemm(false);
+    config.set_debug_options(opts);
+    config.set_num_partitions(num_partitions);
+    TF_ASSIGN_OR_RETURN(auto module,
+                        ParseAndReturnVerifiedModule(hlo_text, config));
+    const int64_t num_params = module->entry_computation()->num_parameters();
+
+    std::vector<std::vector<int64_t>> param_dims(num_params);
+    std::vector<std::vector<int64_t>> param_dims_per_shard(num_params);
+    std::vector<std::vector<int64_t>> param_sharded_dims(num_params);
+    for (int i = 0; i < num_params; ++i) {
+      auto dimensions = module->entry_computation()
+                            ->parameter_instruction(i)
+                            ->shape()
+                            .dimensions();
+      param_dims[i] = std::vector(dimensions.begin(), dimensions.end());
+      param_dims_per_shard[i] = param_dims[i];
+      HloSharding parameter_sharding =
+          module->entry_computation()->parameter_instruction(i)->sharding();
+      EvaluateShardedDims(param_dims_per_shard[i], param_sharded_dims[i],
+                          parameter_sharding);
+    }
+
+    // Slice the tiled inputs to match the prescribed sharding.
+    auto fake_args = xla::MakeFakeArguments(module.get()).value();
```

Comment:
Can you factor out the argument creation into its own function? It's fairly complex and completely separate from the rest of the function.